### PR TITLE
issue-1151 allow to select any correct answer

### PR DESCRIPTION
**Pull request by sarveshwar-gavhane**

### CODE REVIEW

File: js/adapt-contrib-mcq.js [hunk 1]
```diff
@@ -186,6 +186,7 @@ define(function(require) {
             var numberOfRequiredAnswers = 0;
             var numberOfCorrectAnswers = 0;
             var numberOfIncorrectAnswers = 0;
+            var numberOfSelectable=this.model.get('_selectable');
```
Comment (moloko) [inline js/adapt-contrib-mcq.js]: could you adjust your spacing around the '=' here?

File: js/adapt-contrib-mcq.js [hunk 2]
```diff
@@ -213,7 +214,14 @@ define(function(require) {
 
             // Check if correct answers matches correct items and there are no incorrect selections
             var answeredCorrectly = (numberOfCorrectAnswers === numberOfRequiredAnswers) && (numberOfIncorrectAnswers === 0);
-            return answeredCorrectly;
+            if(answeredCorrectly) {
+                return answeredCorrectly;
+            }
+            else {
+               // Check if _selectable matches or less than correct items and there are no incorrect selections
+                var canBeCorrect=(numberOfSelectable <= numberOfCorrectAnswers) && (numberOfIncorrectAnswers === 0);
```
Comment (moloko) [inline js/adapt-contrib-mcq.js]: spacing here too

**Comment by moloko** [comment]
hi @sarveshwar-gavhane I was really hoping to give this a test this week but just had too much on. I'm on holiday now until Wednesday, I will try and have a look then

**Comment by andrewhosgood** [comment]
Instead of introducing a new property, could we not utilise the existing `_selectable`?

Pseudo code:

```js
if( selectable < availableCorrectAnswers ) {
    return ( correctAnswersSelected === selectable ) && incorrectAnswersSelected === 0;
} else {
    return ( correctAnswersSelected === availableCorrectAnswers ) && incorrectAnswersSelected === 0;
}
```

**Comment by sarveshwar-gavhane** [comment]
this was suggested by @moloko  "suppose there are four items, the first three of which could be correct. you have _selectable: 3 but you want the user to get the question right even if they select just one" 
but above code doesn't work with this case.

**Comment by oliverfoster** [comment]
fair point

**Comment by oliverfoster** [comment]
+1

File: js/adapt-contrib-mcq.js [hunk 3]
```diff
@@ -213,7 +213,13 @@ define(function(require) {
 
             // Check if correct answers matches correct items and there are no incorrect selections
             var answeredCorrectly = (numberOfCorrectAnswers === numberOfRequiredAnswers) && (numberOfIncorrectAnswers === 0);
-            return answeredCorrectly;
+            if(answeredCorrectly) {
+                return answeredCorrectly;
+            }
+            else {
+               //Check if _requiredCorrectAnswers is less than or matches numberOfCorrectAnswers and there are no incorrect selections
```
Comment (oliverfoster) [inline js/adapt-contrib-mcq.js]: we tend to do:
```javascript
if (
```
```javascript
} else {
```

**Comment by andrewhosgood** [comment]
Surely if there were 3 right answers and you can select 3 but only select 1, you should get it partly correct?

File: js/adapt-contrib-mcq.js [hunk 4]
```diff
@@ -213,7 +213,12 @@ define(function(require) {
 
             // Check if correct answers matches correct items and there are no incorrect selections
             var answeredCorrectly = (numberOfCorrectAnswers === numberOfRequiredAnswers) && (numberOfIncorrectAnswers === 0);
-            return answeredCorrectly;
+            if(answeredCorrectly) {
+                return answeredCorrectly;
+            } else {
+               //Check if _requiredCorrectAnswers is less than or matches numberOfCorrectAnswers and there are no incorrect selections
+               return (this.model.get('_requiredCorrectAnswers') <= numberOfCorrectAnswers) && (numberOfIncorrectAnswers === 0);
```
Comment (oliverfoster) [inline js/adapt-contrib-mcq.js]: @andrewhosgood makes a good point, this changes the default behaviour of the mcq.
can we:
```javascript
if (answeredCorrectly) {
   // Correct
   return true;
}
if (this.model.get('_requiredCorrectAnswers') == -1) {
   //Incorrect with no specified _requiredCorrectAnswers, assumes _selectable 
   return false;
}
//Check if _requiredCorrectAnswers is less than or matches numberOfCorrectAnswers and there are no incorrect selections
return (this.model.get('_requiredCorrectAnswers') <= numberOfCorrectAnswers) && (numberOfIncorrectAnswers === 0);
```

with default = -1 ?

**Comment by sarveshwar-gavhane** [comment]
with default = -1 gives unexpected behaviour, suppose there are four items, the first three of which could be correct. you have _selectable: 2 and _requiredCorrectAnswers is -1 default , in this case user always gets a partly correct feedback, user never gets correct feedback.
if _selectable and _shouldBeSelected are equal then only  _requiredCorrectAnswers : -1 works fine.
but if _requiredCorrectAnswers is 1 default  its working as expected in all cases.  please correct me if i am wrong.

**Comment by sarveshwar-gavhane** [comment]
@moloko please have a look.

**Comment by moloko** [comment]
@sarveshwar-gavhane I will do I'm just catching up on project work following a few days' holiday

**Comment by moloko** [comment]
> Surely if there were 3 right answers and you can select 3 but only select 1, you should get it partly correct?

This is how MCQ works at the moment. What this PR is trying to do is address the issue you logged in [#1151](https://github.com/adaptlearning/adapt_framework/issues/1151) which is to change that behaviour so that selecting ANY correct answer will result in correct feedback. Or have I misunderstood?

**Comment by moloko** [comment]
It's unfortunate that the property to flag an item is correct is `_shouldBeSelected` - rather than `_isCorrect` which I think would have made all this a bit clearer

**Comment by moloko** [comment]
OK so just working through a few scenarios here:

Scenario 1 (current behaviour): `_selectable` _matches (or exceeds)_ the number of `_shouldBeSelected` items. `_requiredCorrectAnswers` is not present. User must therefore select ALL correct answers to get totally correct feedback, anything less is partly or totally incorrect. This works.

Scenario 2 (@andrewhosgood's suggested change): `_selectable` is _less_ than the number of `_shouldBeSelected` items. The requirement is for the user to select `_selectable` number of correct answers in order get the question totally correct. This can be achieved by setting `_requiredCorrectAnswers` to the same value as `_selectable`. This works.

Scenario 3 (mine & @sarveshwar-gavhane's suggestion): `_selectable` is _less or the same as_ the number of `_shouldBeSelected` items - but you don't want the user to have to select all the `_selectable` items. In this instance you can completely control the minimum number of correct answers required for a totally correct answer by using `_requiredCorrectAnswers`. For example let's say that all the options could be considered correct and therefore all are `_selectable` - but you only require that the user gets two or more correct. This can now be achieved by setting`_requiredCorrectAnswers` to 2.

Perhaps this third scenario is unnecessary - I just figured that if we're adding flexibility into this we may as well go all the way with it!

The only thing is that I can't really see any difference between setting `_requiredCorrectAnswers` to -1 or 1. I am missing something @sarveshwar-gavhane? If there isn't a difference that could work to our advantage as the authoring tool I think will include the `_requiredCorrectAnswers` property by default so it would be good to have something it can be set to by default that will then cause it to be ignored completely.

**Comment by andrewhosgood** [comment]
It's up to you about the third scenario - as long as we come up with a solution to the issue, I'm happy :)

**Comment by sarveshwar-gavhane** [comment]
```
var answeredCorrectly = (numberOfCorrectAnswers === numberOfRequiredAnswers) && (numberOfIncorrectAnswers === 0);
             
            if(answeredCorrectly) {
                return answeredCorrectly;
            }
            if (this.model.get('_requiredCorrectAnswers') < 0 ) {
                    //Incorrect with no specified _requiredCorrectAnswers, assumes _selectable 
                return false;
            }
                //Check if _requiredCorrectAnswers is less than or matches numberOfCorrectAnswers and there are no incorrect selections
               return (this.model.get('_requiredCorrectAnswers') <= numberOfCorrectAnswers) && (numberOfIncorrectAnswers === 0);
```
if answeredCorrectly is false and _requiredCorrectAnswers is by default -1 then above code will always return false, which will give party correct feedback for Scenario 2 and 3

**Comment by sarveshwar-gavhane** [comment]
@moloko  any update on this issue . what can be done ?

**Comment by andrewhosgood** [comment]
I am going to try and have a look and make my own PR based on how I think we have all agreed it should work.

**Comment by andrewhosgood** [comment]
Alternative PR here: https://github.com/adaptlearning/adapt-contrib-mcq/pull/130

I believe that if this works in the way everyone imagined, it is far simpler to use existing variables (`_selectable`) in the logic rather than introduce new ones.

**Comment by moloko** [comment]
@andrewhosgood I'll have a look shortly. I still think it's less flexible but it sounds like I'm the only person who's bothered about that, so I'll shut up about it ;-)

**Comment by sarveshwar-gavhane** [comment]
I have tested PR :  #130 with all three scenarios, it works fine with first two scenarios but it fails for third one. If third scenario is not useful ,then I am okay with PR :  #130

**Comment by andrewhosgood** [comment]
In that case, what about a new parameter for something like `_markAsCorrectIfAtLeastOneCorrectAnswer`? (name is too long, but very accurate)

**Comment by chetan-hajare** [comment]
_markAsCorrectIfAtLeastOneCorrectAnswer sounds good but this will take boolean value(I guess) and this will force user to mark it as correct for one correct answer but the user may wants atleast 2 or 3 correct answers to be selected out of 4 _selectable. For me _requiredCorrectAnswers or _markAsCorrectIfAtLeastCorrectAnswer (this will take number as value) will work in all cases.

**Comment by brian-learningpool** [comment]
I think an optional `_requiredCorrectAnswers` property here sounds good.  As @chetan-hajare mentioned, this offers more flexibility.

**Comment by moloko** [comment]
right I think we have a general consensus that we'd like to move forward with an optional `_requiredCorrectAnswers` property

how far off that is this PR?

**Comment by andrewhosgood** [comment]
In that case, may I suggest if `_requiredCorrectAnswers` is set, we use it else if not, it reverts to the logic in https://github.com/adaptlearning/adapt-contrib-mcq/pull/130/files#diff-fe2ce68e167ce547cebaff3e248a60f5R214?

**Comment by brian-learningpool** [comment]
@oliverfoster has a comment about setting `_requiredCorrectAnswers` to a default of -1 to bypass this activity and maintain the current behaviour for existing configurations.

**Pull request closed without merging**